src/pipelines/data_gen.py

```diff
@@ -166,12 +174,98 @@ def w2v(doc, corpus, w2v_model, w2v, feature):
+        corpusw2v = run_w2v_elemwise(w2v_model, corpus, w2v, operation)
+        vectors = [docw2v,corpusw2v]
+        feature = gen_feature(vectors, w2v, feature)
+
     return feature
 
+def run_w2v_elemwise(w2v_model, doc, w2v, operation):
+    '''
+      Calculates Word2Vec vectors for a document using the first and last sentences of the document
+      Examines vector elements and retains maximum, minimum or absolute value for each vector element
+
+      Args:
+          w2v_model (gensim.Word2Vec): Trained Word2Vec model
+          doc (str): the text of the document
+          w2v (dict): Dictionary of Word2Vec parameters as set in master_pipeline. The dictionary
+           will include keys for the model building parameters min_count, window, size, workers and pretrained.
+           The dict may also have optional boolean keys for the feature operations append, difference, product and cos.
+          operation (str): element wise operation of max, min or abs
+      Returns:
+          documentvector (list): Word2Vec vectors with min/max/abs element values for a sentence, which are then
+          concatenated across sentences
+      '''
+    # Get first and last sentences of document, break down sentences into words and remove stop words
+
+    sentences = get_first_and_last_sentence(doc)
+    normalizedsentences = []
+
+    for sentence in sentences:
+        words = normalize.remove_stop_words(tokenize.word_punct_tokens(sentence))
+        normalizedsentences.append(words)
+
+    sentencevectorarray = []
+
+    # Look up word vectors in trained Word2Vec model and build array of word vectors and sentence vectors
+    for phrase in normalizedsentences:
+
+        # Set up comparison vector based on requested operation
+        if operation == 'max':
+            vectorlist = np.full(w2v['size'], -np.inf)
+        elif operation == 'min':
+            vectorlist = np.full(w2v['size'], np.inf)
+        elif operation == 'abs':
+            vectorlist = np.zeros(w2v['size'])
+
+        # Determine word vector and evaluate elements against comparison vector
+        for word in phrase:
+            try:
+                wordvector = w2v_model[word]
+            except KeyError:
+                continue
+            if operation == 'max':
+                for i in range(0,w2v['size']):
+                    if wordvector[i] > vectorlist[i]:
+                        vectorlist[i] = wordvector[i]
+            elif operation == 'min':
+                for j in range(0, w2v['size']):
+                    if wordvector[j] < vectorlist[j]:
+                        vectorlist[j] = wordvector[j]
+            elif operation == 'abs':
+                for k in range(0, w2v['size']):
+                    if abs(wordvector[k]) > vectorlist[k]:
+                        vectorlist[k] = abs(wordvector[k])
+
+        # Remove any infinity values from special cases (ex: 1 word sentence and word not in word2vec model)
+        if any(np.isinf(vectorlist)):
```

Comment:
on a similar note I think this block can be expressed as `vectorlist = np.where(np.isinf(vectorlist), 0, vectorlist)`